05-mixed-poisson.ipynb

```diff
@@ -1,519 +1,608 @@
+        "\n",
+        "# Load Firedrake on Colab\n",
+        "try:\n",
+        "    import firedrake\n",
+        "except ImportError:\n",
+        "    !wget \"https://fem-on-colab.github.io/releases/firedrake-install-real.sh\" -O \"/tmp/firedrake-install.sh\" && bash \"/tmp/firedrake-install.sh\"\n",
+        "    import firedrake\n",
+        "\n",
+        "from firedrake import *"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 12,
+      "metadata": {
+        "id": "jmZ4mZv1vXp_"
+      },
+      "outputs": [],
+      "source": [
+        "from time import time\n",
+        "\n",
+        "N = 20\n",
+        "\n",
+        "mesh = UnitSquareMesh(N, N)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "CwjuKh6dvXqA"
+      },
+      "source": [
+        "We need to choose an LBB-stable finite element pair for $W = V\\times\\Sigma$, one choice is the Raviart-Thomas space of degree $k$ for $\\Sigma$ and discontinuous polynomials of degree $k$ for $V$, here we choose $k = 1$."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 201
+        },
+        "id": "EyVXY8lGvXqA",
+        "outputId": "db1de22f-939d-4e9c-806c-ed1e12bc2d9c"
+      },
+      "outputs": [
+        {
+          "output_type": "error",
+          "ename": "NameError",
+          "evalue": "ignored",
+          "traceback": [
+            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
+            "\u001b[0;32m<ipython-input-5-bb5f3b5e2438>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFiniteElement\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Raviart-Thomas\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtriangle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"integral\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mSigma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFunctionSpace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmesh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mV\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFunctionSpace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmesh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"DG\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
+            "\u001b[0;31mNameError\u001b[0m: name 'FiniteElement' is not defined"
```

Comment:
Hm it looks like something is broken?

Reply:
Ah sorry Joe that would also be fixed by 'from firedrake import *' 

I'll tidy up the outputs